node/impl/full/eth_events.go

```diff
@@ -341,6 +343,35 @@ func (e *ethSubscription) start(ctx context.Context) {
 					e.send(ctx, r)
 				}
 			case *types.TipSet:
+				// Get current chain epoch
+				head, err := e.ChainAPI.ChainHead(ctx)
```

Comment:
Unfortunately, anything that looks at the head here is going to be racy because there could have been a reorg in the meantime

I think we can just return the tipset's parent here. Although we probably need to remember the last tip set we returned, so we avoid returning it multiple times in a row (e.g., if the "pending" tipset changes).

Reply:
@Stebalien Please can you explain the race here in more detail ? Even if the head changes by the time we get here, the epoch should still remain the same right ? I am interested in the epoch of the head here and not the head itself.

Reply:
I don't think that's guaranteed due to shenanigans with null tipsets and weights. Finally, all you need is a heavier chain (more blocks). If I recall correctly, height is a factor in weight, but there are a bunch of other factors as well.

Regardless, it's simpler to just return the parent here. It means we don't have to keep some kind of buffer or anything.

Reply:
Makes sense.